parmed/modeller/residue.py

```diff
@@ -432,25 +436,25 @@ def apply_patch(self, patch, precision=4):
                         residue.tail = None
                 # Add bond
                 residue.add_bond(atom1_name, atom2_name, order)
-            except:
-                raise Exception('Bond %s-%s could not be added to patched residue: atoms are %s' % (atom1_name, atom2_name, list(residue._map.keys())))
+            except IndexError as e:
+                raise IncompatiblePatchError('Bond %s-%s could not be added to patched residue: atoms are %s' % (atom1_name, atom2_name, list(residue._map.keys())))
         # Delete impropers
         for impr in patch.delete_impropers:
             try:
                 residue.remove(impr)
-            except:
-                raise Exception('Improper %s was not found in residue to be patched.' % impr)
+            except ValueError as e:
+                raise IncompatiblePatchError('Improper %s was not found in residue to be patched.' % impr)
```

Comment:
I am not why we need to create this. Do we need to catch it elsewhere?

Reply:
I was instructed by @swails to create a new exception type that is as narrow as possible for users of this code to be able to catch when the patch they are attempting to apply is not compatible. If successful, the method returns the patched `ResidueTemplate`. If unsuccessful, the method raises `IncompatiblePatchError`. This is easy to catch if one wants to check whether the patch is incompatible, as we do.

Reply:
Specifically, it's caught in `parmed/openmm/parameters.py` in `_determine_valid_patch_combinations()`. Again, @swails preferred we catch *only* the intended exceptions.

Reply:
My current understanding is that the only way to see if a patch applies (without duplicating logic here) is to try it and catch any exceptions.  It would be useful, then, to catch an exception that indicated *specifically* this purpose, rather than rely on some kind of built-in exception that has other uses throughout Python.

ParmEd is complex (moreso than it needed to be, but every MD program has enough quirks to make simplicity difficult to achieve), and there are a lot of ways of "breaking" things (Python doesn't impose type safety, and ParmEd simply assumes compatible types wherever they're used in data structures that ParmEd populates with various parsers).  So catching general exceptions like `KeyError` and `IndexError` to identify incompatible patches has a non-negligible likelihood of masking unrelated errors.  It's best to fail as quickly as possible to make debugging as easy as possible, and this new exception furthers that goal.

Reply:
@ all: I see.